### PR TITLE
feat: add applock override through MDM [WPB-26671]

### PR DESCRIPTION
**Pull request by e-maad**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-26671" title="WPB-26671" target="_blank">TaskWPB-26671</a>  [Web] Bypass Applock using MDM config for Managed devices
  </td></table>
  <br />
 

### Description

Allow app to consume MDM config and influence app lock functionality.
There is an env veraible to control the behaviour

### CODE REVIEW

File: apps/webapp/src/script/repositories/user/appLockState.ts [hunk 1]
```diff
@@ -63,6 +64,14 @@ export class AppLockState {
     this.isAppLockActivated = ko.pureComputed(() => this.isAppLockEnabled() && this.hasPassphrase());
     this.hasPassphrase = ko.observable(false);
     this.isActivatedInPreferences = ko.observable(false);
-    this.isAppLockEnabled = ko.pureComputed(() => this.isAppLockEnforced() || this.isActivatedInPreferences());
+    this.isAppLockEnabled = ko.pureComputed(() => {
```
Comment (zskhan) [inline apps/webapp/src/script/repositories/user/appLockState.ts:67]: Can we add tests to ensure the expectation?

File: apps/webapp/src/script/page/mainContent/panels/preferences/accountPreferences/privacySection.tsx [hunk 2]
```diff
@@ -112,15 +117,14 @@ const PrivacySection = ({
           {translate('preferencesAccountTypingIndicatorsDetail')}
         </p>
       </div>
-
       {isAppLockAvailable && (
```
Comment (zskhan) [inline apps/webapp/src/script/page/mainContent/panels/preferences/accountPreferences/privacySection.tsx:120]: Also add few tests to verify the visibility of checkbox

File: apps/webapp/src/script/page/mainContent/panels/preferences/accountPreferences/privacySection.tsx [hunk 3]
```diff
@@ -54,6 +55,10 @@ const PrivacySection = ({
       'appLockInactivityTimeoutSecs',
     ]);
 
+  const enableMdmConfig = Config.getConfig().FEATURE.ENABLE_MDM_CONFIG;
+  const appLockOverride = Boolean(Config.getDesktopConfig()?.managedConfig?.applockOverride);
```
Comment (arjita-mitra) [inline apps/webapp/src/script/page/mainContent/panels/preferences/accountPreferences/privacySection.tsx]: Boolean(value) silently accepts any truthy/falsy value. The strict comparison === true enforces that contract.

```
// This will violate your contract but pass lint:
window.desktopAppConfig = {
  managedConfig: {
    applockOverride: 1  // Number, not boolean
  }
};
```

`const appLockOverride = Boolean(Config.getDesktopConfig()?.managedConfig?.applockOverride);`
this will pass because 1 is truthy

```
const appLockOverride = Boolean(
  Config.getDesktopConfig()?.managedConfig?.applockOverride === true
);
```
this will correctly reject as 1 === true is false.

Recommended

```
const appLockOverride = Boolean(
  Config.getDesktopConfig()?.managedConfig?.applockOverride === true
);
```

**Review by arjita-mitra — CHANGES_REQUESTED** [review]

File: apps/webapp/src/script/repositories/user/appLockState.ts [hunk 4]
```diff
@@ -63,6 +64,14 @@ export class AppLockState {
...
+      const isMdmAppLockDisabled =
+        Config.getConfig().FEATURE.ENABLE_MDM_CONFIG &&
```
Comment (arjita-mitra) [inline apps/webapp/src/script/repositories/user/appLockState.ts:69]: Feedback for UX/PM. When MDM disables app lock:

- The UI checkbox is disabled
- But there's no explanation to the user why it's disabled?
- Users might be confused why they can't toggle app lock.

we could add a simple help text block

```
{isMdmAppLockDisabled && (
  <p>
    {translate('preferencesAccountAppLockDisabledByMDM')}
  </p>
)}
```

File: apps/webapp/src/script/page/mainContent/panels/preferences/accountPreferences/privacySection.tsx [hunk 5]
```diff
@@ -112,15 +117,14 @@ const PrivacySection = ({
...
-
       {isAppLockAvailable && (
         <div className="checkbox-margin">
           <Checkbox
             onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
               appLockRepository.setEnabled(event.target.checked);
             }}
             checked={isAppLockEnabled}
-            disabled={isAppLockEnforced}
+            disabled={isAppLockEnforced || isMdmAppLockDisabled}
```
Comment (arjita-mitra) [inline apps/webapp/src/script/page/mainContent/panels/preferences/accountPreferences/privacySection.tsx:127]: No Tests Added

MDM override when feature flag is enabled
MDM override when feature flag is disabled
Missing managed config object
Edge cases with stale Desktop config etc

File: apps/webapp/src/script/repositories/user/appLockState.ts (hunk 4 above)
Comment (e-maad) [reply apps/webapp/src/script/repositories/user/appLockState.ts:69]: Good point, this is just for the first iteration and it will be behind FF.
I will discuss this for the next iteration.

File: apps/webapp/src/script/page/mainContent/panels/preferences/accountPreferences/privacySection.test.tsx [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
+    isActivatedInPreferences,
+    isAppLockDisabledOnTeam: ko.pureComputed(() => isAppLockDisabledOnTeamValue()),
+  };
+};
+
+const createMockPropertiesRepository = (): PropertiesRepository => {
+  return {
+    receiptMode: ko.observable(0),
+    typingIndicatorMode: ko.observable(0),
+  } as any;
```
Comment (zskhan) [inline apps/webapp/src/script/page/mainContent/panels/preferences/accountPreferences/privacySection.test.tsx]: No any please

File: apps/webapp/src/script/page/mainContent/panels/preferences/accountPreferences/privacySection.test.tsx [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+};
+
+const createMockAppLockRepository = (): AppLockRepository => {
+  return {
+    setEnabled: jest.fn(),
+  } as any;
```
Comment (zskhan) [inline apps/webapp/src/script/page/mainContent/panels/preferences/accountPreferences/privacySection.test.tsx]: No any please

File: apps/webapp/src/script/page/mainContent/panels/preferences/accountPreferences/privacySection.test.tsx [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+describe('PrivacySection', () => {
+  let configSpy: jest.SpyInstance;
+  let desktopConfigSpy: jest.SpyInstance;
+
+  beforeEach(() => {
+    configSpy = jest.spyOn(Config, 'getConfig').mockReturnValue({
+      FEATURE: {
+        ENABLE_MDM_CONFIG: false,
+      },
+    } as any);
```
Comment (zskhan) [inline apps/webapp/src/script/page/mainContent/panels/preferences/accountPreferences/privacySection.test.tsx]: No any please

**Review by zskhan — CHANGES_REQUESTED** [review]

**Review by arjita-mitra — APPROVED** [review]

**Review by zskhan — APPROVED** [review]
